### ISSUE TITLE
Go indexer: Expand the treatment of MarkedSource. [phab:D1539]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1957]
_Created by **fromberger** at 2017-04-19 05:29:36_: 

For variable, type, and function declarations, provide enough context for the
renderer to extract identifiers and basic type information.

Types are somewhat abbreviated, as the interior structure of struct and
interface declarations is generally not that interesting in the UI without the
code present -- we may want to revisit that later if we start doing more
elaborate things with the structure.

Add test cases for code facts.

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-04-19 05:31:02_:

  - Remove now-unused smoke.go test file.

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-04-19 16:03:29_:

  - Fix an incomplete

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-04-20 18:10:29_:

MarkedSource for type nodes should not be dominated by TYPE (otherwise, if we try to print the simple identifier/qualified name without including type information, we'd get nothing back).

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-04-20 18:10:29_:

What about x's type?

Can you use PARAMETER_LOOKUP_BY_PARAM for functions to reduce the amount of duplicate data?

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-04-20 19:25:49_:

I originally tried that, and wasn't able to get it to work satisfactorily. 
I'll bash on it some more.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-04-20 19:25:49_:

Apparently I'm holding it wrong then. I tried rendering these as a plain BOX, but then the doc tool didn't render anything at all except the identifier, even in the doc view.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-04-20 19:49:48_:

The doc tool won't issue the query (since it's only rendering the MarkedSource it gets as input). The serving infrastructure converts PARAMETER_LOOKUP_BY_PARAM to the parameter list.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-04-20 19:49:48_:

```
  marked_source { # ThingerCode
    child { }
    child { # THC1
      kind: BOX
      child { # THC1Context
        kind: CONTEXT
        child { # THC1ContextID
          kind: IDENTIFIER
          pre_text: "iface"
        }
        post_child_text: "."
        add_final_list_token: true
      }
      child { # THC1Ident
        kind: IDENTIFIER
        pre_text: "Thinger"
      }
    }
    child { # THC2
      pre_text: " "
    }
    child { # THC3
      pre_text: "interface {...}"
    }
  }
```
renders properly for me in the doc tool (though the `interface {...}` bit doesn't show since it's not part of what is considered interesting for doc snippets).

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-04-21 15:59:07_:

  - Remove now-unused smoke.go test file.
  - Fix an incomplete
  - Do not dominate identifiers by TYPE marks.

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-04-21 16:09:13_:

  - Use PARAMETER_LOOKUP_BY_PARAM for function/method parameters.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-04-21 16:12:20_:

I guess I'll try it and see what it does. It's easy enough to tweak it if I get it wrong. Done. :)

**Comment by kythe** [comment]
_**zarko** accepted this revision at 2017-04-24 16:56:37_

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-04-24 17:04:10_:

Thanks for your helpful comments!

**Comment by kythe** [comment]
_**fromberger** closed via commit 44e0c754fe0ab9f8674f16ffd7e332827cbaf50a at 2017-04-24 17:04:36_
